### PR TITLE
(BKR-1155) Allow custom ssh connection preference

### PR DESCRIPTION
**Pull request by rishijavia**

### CODE REVIEW

**Comment by rishijavia** [comment]
original [pr](https://github.com/puppetlabs/beaker/pull/1442) submitted to beaker

**Comment by kevpl** [comment]
@puppetlabs-jenkins test this please

**Comment by rishijavia** [comment]
@kevpl done! :+1:

**Comment by kevpl** [comment]
@puppetlabs-jenkins test this please

**Comment by kevpl** [comment]
@puppetlabs-jenkins retest this please

**Comment by kevpl** [comment]
@puppetlabs-jenkins retest this please

File: lib/beaker/hypervisor.rb [hunk 1]
```diff
@@ -57,6 +58,26 @@ def cleanup
...
 
+    DEFAULT_CONNECTION_PREFERENCE = [:ip, :vmhostname, :hostname]
+    # SSH connection method preference. Can be overwritten by hypervisor to change the order
+    def connection_preference(host)
+      DEFAULT_CONNECTION_PREFERENCE
+    end
+
+    def self.set_ssh_connection_preference(hosts_to_provision, hypervisor)
+      hosts_to_provision.each do |host|
+        ssh_methods = hypervisor.connection_preference(host)
```
Comment (ferglor) [inline lib/beaker/hypervisor.rb]: is it cleaner to change this line to `ssh_methods = hypervisor.connection_preference(host) + DEFAULT_CONNECTION_PREFERENCE`

File: lib/beaker/hypervisor.rb [hunk 2]
```diff
@@ -57,6 +58,26 @@ def cleanup
...
+        if host[:ssh_preference]
+          # If user has provided ssh_connection_preference in hosts file then concat the preference provided by hypervisor
+          # Followed by concatenating the default preference and keeping the unique once
+          ssh_methods = (host[:ssh_preference] + (ssh_methods + DEFAULT_CONNECTION_PREFERENCE)).uniq
```
Comment (ferglor) [inline lib/beaker/hypervisor.rb]: ...and then change this line to `ssh_methods = host[:ssh_preference] + ssh_methods if host[:ssh_preference]`

File: lib/beaker/hypervisor.rb [hunk 3]
```diff
@@ -57,6 +58,26 @@ def cleanup
...
+        else
+          ssh_methods = (ssh_methods + DEFAULT_CONNECTION_PREFERENCE).uniq
```
Comment (ferglor) [inline lib/beaker/hypervisor.rb]: ...and then get rid of this else block

File: lib/beaker/hypervisor.rb [hunk 4]
```diff
@@ -57,6 +58,26 @@ def cleanup
...
+        end
+        host[:ssh_connection_preference] = ssh_methods
```
Comment (ferglor) [inline lib/beaker/hypervisor.rb]: ...and then change this line to `host[:ssh_connection_preference] = ssh_methods.uniq`

**Comment by ferglor** [comment]
@puppetlabs-jenkins retest this please

File: lib/beaker/hypervisor.rb (hunk 4 above)
Comment (rishijavia) [inline lib/beaker/hypervisor.rb]: @ferglor you're right this definitely looks cleaner. I'll update this soon

**Comment by kevpl** [comment]
@ferglor did @rishijavia get everything you were thinking of? If so, then you should re-kick testing, it looks like the failure was unrelated as far as I'm seeing.

**Comment by ferglor** [comment]
@puppetlabs-jenkins retest this please

**Comment by ferglor** [comment]
@puppetlabs-jenkins retest this please

**Comment by kevpl** [comment]
@puppetlabs-jenkins retest this please

**Comment by kevpl** [comment]
@puppetlabs-jenkins retest this please

**Pull request merged**